pkg/cmd/infra/deployer/deployer.go

```diff
@@ -77,88 +74,111 @@ func NewCommandDeployer(name string) *cobra.Command {
+	if !hasDesired {
+		return fmt.Errorf("deployment %s has no desired replica count")
 	}
 
-	if newConfig, err = deployutil.DecodeDeploymentConfig(newDeployment, latest.Codec); err != nil {
-		return nil, nil, err
+	// Find all controllers in order to pick out the deployments.
+	controllers, err := d.getControllers(namespace)
+	if err != nil {
+		return fmt.Errorf("couldn't get controllers in namespace %s: %v", namespace, err)
 	}
 
-	glog.Infof("Found new Deployment %s for DeploymentConfig %s/%s with latestVersion %d", newDeployment.Name, newConfig.Namespace, newConfig.Name, newConfig.LatestVersion)
-
-	// Collect all deployments that predate the new one by comparing all old ReplicationControllers with
-	// encoded DeploymentConfigs to the new one by LatestVersion. Treat a failure to interpret a given
-	// old deployment as a fatal error to prevent overlapping deployments.
-	var allControllers *kapi.ReplicationControllerList
-	oldDeployments := []*kapi.ReplicationController{}
+	// Find all deployments sorted by version.
+	deployments := deployutil.ConfigSelector(config.Name, controllers.Items)
+	sort.Sort(deployutil.DeploymentsByLatestVersionDesc(deployments))
 
-	if allControllers, err = controllerGetter.List(newDeployment.Namespace, labels.Everything()); err != nil {
-		return nil, nil, fmt.Errorf("unable to get list replication controllers in deployment namespace %s: %v", newDeployment.Namespace, err)
+	// Find any last completed deployment.
+	var lastDeployment *kapi.ReplicationController
+	for _, candidate := range deployments {
+		if candidate.Name == deployment.Name {
+			continue
+		}
+		if deployutil.DeploymentStatusFor(&candidate) == deployapi.DeploymentStatusComplete {
+			lastDeployment = &candidate
+			glog.Infof("Picked %s as the last completed deployment", deployutil.LabelForDeployment(&candidate))
+			break
+		}
+	}
+	if lastDeployment == nil {
+		glog.Info("No last completed deployment found")
 	}
 
-	glog.Infof("Inspecting %d potential prior deployments", len(allControllers.Items))
-	for i, controller := range allControllers.Items {
-		if oldName := deployutil.DeploymentConfigNameFor(&controller); oldName != newConfig.Name {
-			glog.Infof("Disregarding deployment %s (doesn't match target DeploymentConfig %s)", controller.Name, oldName)
+	// Scale down any deployments which aren't the new or last deployment.
+	for _, candidate := range deployments {
+		if candidate.Name == deployment.Name {
 			continue
 		}
-
-		if deployutil.DeploymentVersionFor(&controller) < newConfig.LatestVersion {
-			glog.Infof("Marking deployment %s as a prior deployment", controller.Name)
-			oldDeployments = append(oldDeployments, &allControllers.Items[i])
+		if lastDeployment != nil && candidate.Name == lastDeployment.Name {
+			continue
+		}
+		if candidate.Spec.Replicas == 0 {
+			// The deployment is already scaled down
+			continue
+		}
+		candidate.Spec.Replicas = 0
+		_, err := d.updateDeployment(namespace, &candidate)
```

Comment:
Well, I made the first comment because the glog you were printing as soon as you updated the deployment was informing the user that it was scaled which is not true ("scaling" would be more precise). I still believe that not waiting for those deployments is fine and we probably shouldn't...